File: lightbulb-recycleradapters/src/main/java/com/github/RooneyAndShadows/lightbulb/recycleradapters/implementation/RadioButtonSelectableAdapter.java
package com.github.rooneyandshadows.lightbulb.recycleradapters.implementation;

import android.graphics.drawable.Drawable;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.github.rooneyandshadows.lightbulb.recycleradapters.R;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterConfiguration;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterDataModel;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterSelectableModes;
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyRecyclerAdapter;
import com.github.rooneyandshadows.lightbulb.selectableview.RadioButtonView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

@SuppressWarnings({"unchecked"})
public class RadioButtonSelectableAdapter<ItemType extends EasyAdapterDataModel> extends EasyRecyclerAdapter<ItemType> {

    public RadioButtonSelectableAdapter() {
        super(new EasyAdapterConfiguration<ItemType>().withSelectMode(EasyAdapterSelectableModes.SELECT_SINGLE));
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        RadioButtonView v = (RadioButtonView) LayoutInflater.from(parent.getContext())
                .inflate(R.layout.layout_radio_button, parent, false);
        int[] padding = getItemPadding();
        if (padding != null && padding.length == 4)
            v.setPadding(padding[0], padding[1], padding[2], padding[3]);
        return new RadioButtonViewHolder(v);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        RadioButtonViewHolder vHolder = (RadioButtonViewHolder) holder;
        vHolder.bindItem();
    }

    @Override
    public void onViewRecycled(@NonNull RecyclerView.ViewHolder holder) {
        RadioButtonViewHolder vh = (RadioButtonViewHolder) holder;
        vh.recycle();
    }

    @Override
    public int getItemCount() {
        return items.size();
    }

    @Override
    public String getItemName(ItemType item) {
        return item.getItemName();
    }

    protected Drawable getItemIcon(ItemType item) {
        return null;
    }

    protected Drawable getItemIconBackground(ItemType item) {
        return null;
    }

    /**
     * Defines padding for the item view.
     *
     * @return int[] {left,top,right,bottom}
     */
    protected int[] getItemPadding() {
        return null;
    }

    public class RadioButtonViewHolder extends RecyclerView.ViewHolder {
        protected RadioButtonView selectableView;
        protected ItemType item;

        public RadioButtonViewHolder(RadioButtonView categoryItemBinding) {
            super(categoryItemBinding);
            this.selectableView = (RadioButtonView) itemView;
            selectableView.setOnCheckedListener((view, isChecked) -> selectableView.post(() -> selectItem(item, isChecked)));
        }

        public void bindItem() {
            this.item = getItem(getBindingAdapterPosition());
            boolean isSelectedInAdapter = isItemSelected(item);
            if (selectableView.isChecked() != isSelectedInAdapter)
                selectableView.setChecked(isSelectedInAdapter);
            selectableView.setText(getItemName(item));
            selectableView.setIcon(getItemIcon(item), getItemIconBackground(item));
        }

        private void recycle() {
            selectableView.setIcon(null, null);
        }
    }
}

File: lightbulb-recycleradapters/src/main/java/com/github/RooneyAndShadows/lightbulb/recycleradapters/implementation/RadioButtonSelectableAdapter.kt
package com.github.rooneyandshadows.lightbulb.recycleradapters.implementation

import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import com.github.rooneyandshadows.lightbulb.recycleradapters.R
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterDataModel
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyAdapterSelectableModes
import com.github.rooneyandshadows.lightbulb.recycleradapters.abstraction.EasyRecyclerAdapter
import com.github.rooneyandshadows.lightbulb.selectableview.RadioButtonView

@Suppress("UNUSED_PARAMETER", "unused", "MemberVisibilityCanBePrivate")
class RadioButtonSelectableAdapter<ItemType : EasyAdapterDataModel> :
    EasyRecyclerAdapter<ItemType>(EasyAdapterSelectableModes.SELECT_SINGLE) {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        val v = LayoutInflater.from(parent.context)
            .inflate(R.layout.layout_radio_button, parent, false) as RadioButtonView
        val padding = itemPadding
        if (padding != null && padding.size == 4) v.setPadding(
            padding[0],
            padding[1],
            padding[2],
            padding[3]
        )
        return RadioButtonViewHolder(v)
    }

    @Suppress("UNCHECKED_CAST")
    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        val vHolder: RadioButtonViewHolder = holder as RadioButtonSelectableAdapter<ItemType>.RadioButtonViewHolder
        vHolder.bindItem()
    }

    @Suppress("UNCHECKED_CAST")
    override fun onViewRecycled(holder: RecyclerView.ViewHolder) {
        val vh: RadioButtonViewHolder = holder as RadioButtonSelectableAdapter<ItemType>.RadioButtonViewHolder
        vh.recycle()
    }

    override fun getItemCount(): Int {
        return getItems().size
    }

    override fun getItemName(item: ItemType): String {
        return item.itemName
    }

    protected fun getItemIcon(item: ItemType): Drawable? {
        return null
    }

    protected fun getItemIconBackground(item: ItemType): Drawable? {
        return null
    }

    /**
     * Defines padding for the item view.
     *
     * @return int[] {left,top,right,bottom}
     */
    protected val itemPadding: IntArray?
        get() = null

    inner class RadioButtonViewHolder(categoryItemBinding: RadioButtonView?) :
        RecyclerView.ViewHolder(
            categoryItemBinding!!
        ) {
        private var selectableView: RadioButtonView = itemView as RadioButtonView
        private var item: ItemType? = null
        fun bindItem() {
            item = getItem(bindingAdapterPosition) ?: return
            val isSelectedInAdapter = isItemSelected(item!!)
            if (selectableView.isChecked != isSelectedInAdapter) selectableView.isChecked =
                isSelectedInAdapter
            selectableView.text = getItemName(item!!)
            selectableView.setIcon(getItemIcon(item!!), getItemIconBackground(item!!))
        }

        fun recycle() {
            selectableView.setIcon(null, null)
        }

        init {
            selectableView.setOnCheckedListener { _: RadioButtonView?, isChecked: Boolean ->
                selectableView.post {
                    selectItem(
                        item!!,
                        isChecked
                    )
                }
            }
        }
    }
}